src/libopensc/card-cac.c

```diff
@@ -0,0 +1,1540 @@
+			pki_obj.path.aid.value[pki_obj.path.aid.len-1] = i;
+			pki_obj.fd = i+1; /* don't use id of zero */
+			cac_add_object_to_list(&priv->pki_list, &pki_obj);
+		}
+	}
+
+	/* populate non-PKI objects */
+	for (i=0; i < cac_1_object_count; i++) {
+		r = cac_select_file_by_type(card, &cac_1_objects[i].path, NULL, SC_CARD_TYPE_CAC_I);
+		if (r == SC_SUCCESS) {
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,"CAC1: obj_object found, cert_next=%d (%s),", i, cac_1_objects[i].name);
+			cac_add_object_to_list(&priv->general_list, &cac_1_objects[i]);
+		}
+	}
+
+	/*
+	 * create a cuid to simulate the cac 2 cuid.
+	 */
+	priv->cuid = cac_cac1_cuid;
+	/* create a serial number by hashing the first 100 bytes of the
+	 * first certificate on the card */
+	r = cac_select_pki_applet(card, index);
+	if (r < 0) {
+		return r; /* shouldn't happen unless the card has been removed or is malfunctioning */
+	}
+	val = buf;
+	val_len = sizeof(buf);
+	r = cac_cac1_get_certificate(card, &val, &val_len);
+	if (r >= 0) {
+		priv->cac_id = malloc(20);
+		if (priv->cac_id == NULL) {
+			return SC_ERROR_OUT_OF_MEMORY;
+		}
+#ifdef ENABLE_OPENSSL
+		SHA1(val, val_len, priv->cac_id);
+		priv->cac_id_len = 20;
+#else
+		sc_log(card->ctx, "OpenSSL Required");
+		return SC_ERROR_NOT_SUPPORTED;
+#endif /* ENABLE_OPENSSL */
+	}
+	return SC_SUCCESS;
+}
+
+/*
+ * Look for a CAC card. If it exists, initialize our data structures
+ */
+static int cac_find_and_initialize(sc_card_t *card)
+{
+	int r, index;
+	cac_private_data_t *priv = NULL;
+
+	/* already found? */
+	if (card->drv_data) {
+		return SC_SUCCESS;
+	}
+
+	/* is this a CAC-2 specified in NIST Interagency Report 6887 -
+	 * "Government Smart Card Interoperability Specification v2.1 July 2003" */
+	r = cac_select_CCC(card);
+	if (r == SC_SUCCESS) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,"CCC found, is CAC-2");
+		priv = cac_new_private_data();
+		r = cac_process_CCC(card, priv);
+		if (r == SC_SUCCESS) {
+			card->type = SC_CARD_TYPE_CAC_II;
+			card->drv_data = priv;
+			return r;
+		}
+	}
+
+	/* is this a CAC-1 specified in DoD "CAC Applet Developer Guide" version 1.0 September 2002 */
+	r = cac_find_first_pki_applet(card, &index);
+	if (r == SC_SUCCESS) {
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,"applet found, is CAC-1");
+		if (!priv) {
+			priv = cac_new_private_data();
+		}
+		r = cac_populate_cac_1(card, index, priv);
+		if (r == SC_SUCCESS) {
+			card->type = SC_CARD_TYPE_CAC_I;
+			card->drv_data = priv;
+			return r;
+		}
+	}
+	if (priv) {
+		cac_free_private_data(priv);
+	}
+	return r;
+}
+
+
+/* NOTE: returns a bool, 1 card matches, 0 it does not */
+static int cac_match_card(sc_card_t *card)
+{
+	int r;
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+	/* Since we send an APDU, the card's logout function may be called...
+	 * however it may be in dirty memory */
+	card->ops->logout = NULL;
+
+	r = cac_find_and_initialize(card);
```

Comment:
It may be true now that match_card() is only called once. But who knows what will happen in the future. Better keep the implementation similar to what a naive developer would expect ;-)

But you can be sure that `match_card` has been called prior to `init`, see a7a903fd. This means that you could set `card->type` in `match_card` and use this value in `init`.